runtime/src/main/cpp/Exceptions.cpp

```diff
@@ -227,71 +226,103 @@ void OnUnhandledException(KRef throwable) {
   }
 }
 
-#if KONAN_REPORT_BACKTRACE_TO_IOS_CRASH_LOG
-static bool terminating = false;
-static SimpleMutex terminatingMutex;
-#endif
+namespace {
 
-RUNTIME_NORETURN void TerminateWithUnhandledException(KRef throwable) {
-  OnUnhandledException(throwable);
+class {
+    /**
+     * Timeout 5 sec for concurrent (second) terminate attempt to give a chance the first one to finish.
+     * If the terminate handler hangs for 5 sec it is probably fatally broken, so let's do abnormal _Exit in that case.
+     */
+    unsigned int timeoutSec = 5;
+    int terminatingFlag = 0;
+  public:
+    template <class Fun> RUNTIME_NORETURN void operator()(Fun block) {
+      if (!compareAndSet(&terminatingFlag, 0, 1)) {
+        block();
+        // block() is supposed to be NORETURN, otherwise go to normal abort()
+        konan::abort();
+      } else {
+        sleep(timeoutSec);
+        // We come here when another terminate handler hangs for 5 sec, that looks fatally broken. Go to forced exit now.
+      }
+      _Exit(EXIT_FAILURE); // force exit
```

Comment:
I'm afraid one of my questions got lost. What about logging to stderr facts (1) that there's a concurrent termination attempt and (2) that one of them got tired of waiting and is force quitting?

Reply:
This may be dangerous, as output itself is a sort of conspicuous regarding races, etc. I consider hanging termination (which already happens here) as extremely emergence case, something is awfully wrong - it may be heap corruption or whatever that affects any printing. So I'd prefer to minimize any actions.